### ISSUE TITLE
ICE on self-referential typedef

### ISSUE THREAD

**Issue by imbrem** [rust-lang/rust#62364]
So, while writing some _really nasty_ code for an interpreter, I ended up with an overcomplicated version of this code:
```rust
use std::sync::Arc;

pub enum ThingOrRef<S> where S: AsRef<Self> {
    ThingRef(S),
    Thing
}

pub type ThingArena<'ctx, T> = ThingOrRef<Arc<Self>>;
```
([Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=478e073ccd146e94c0244792d167ce08))

Alas, it leads to an ICE, with this message:
```
thread 'rustc' panicked at 'index out of bounds: the len is 1 but the index is 18446744073709551615'
```
I wonder how an unchecked overflow got in _there_ (18446744073709551615 == -1)... 

Funnily enough, if I remove the `'ctx` _or_ the `T` (or both) it compiles just fine, without error. But with both the `ctx` and `T` in, crash.

**Comment by imbrem** [comment]
Note in my original code I used the `'ctx` with a kind of home-grown `Cow`, but it turns out it just needs to be there, not used. So both cases crash. The `T` was also used.

**Comment by Centril** [comment]
Reduced:
```rust
struct Struct<P1> {
    field: P1,
}

type Alias<'l1, P2> = Struct<*const Self>;
```

**Comment by imbrem** [comment]
Would this be a good issue for a beginner such as myself to take a crack at?

**Comment by hellow554** [comment]
Regression in 740668dbd99dbf1726bbb0cca6cd0943ea2f7e27

Therefore probably dup of #62263

**Comment by pnkfelix** [comment]
triage: P-high, removing nomination. Assigning to @alexreg (as well as self to ensure follow-up)

**Comment by alexreg** [comment]
@imbrem Probably not the easiest. :-) I'm going to look at solving this, but maybe take a look at what I do if you like.

**Comment by alexreg** [comment]
Turned out to be far easier than thought, though identifying the solution wasn't so straightforward perhaps... PR fix upcoming.

**Comment by imbrem** [comment]
@alexreg nice! I'll probably have a look when I have more time on the weekend. Also are those Silmarils in your profile picture?

**Comment by alexreg** [comment]
@imbrem Sure. The actual change itself is mainly just removal of a few lines that I'd inserted in a previous PR. Most of the PR will be tests heh.

And yep, the Silmarils with one of the Two Trees. :-)

**Comment by pnkfelix** [comment]
Fixed by #62364

**Comment by lzutao** [comment]
I think it is #62417.
